feat: CRUD Exam

MaksiConGit opened the pull request:
- Se realizó el CRUD de Exam con las validaciones.
- Se creó la ruta
- Se cambió  la seeder, factory y migration de exam para cambiar 'exam_number' por 'number'
- Se cambiaron las validaciones de StoreExamRequest para que valide teacher_id y subject_id por separado
- Store y Update ahora verifican si existe un teacher_subject_id del teacher_id y subject_id ingresados en el formulario, si existe, guarda su id, sino existe, crea un nuevo registro con estos datos y lo guarda.
- Se creó un modelo TeacherSubject el cual está relacionado con Teacher y Subject
- En el modelo de Exam se crearon las relaciones de Teacher y Subject

MaksiConGit committed:
feat: CRUD Exam

- Se realizó el CRUD de Exam con las validaciones.
- Se creó la ruta
- Se cambió  la seeder, factory y migration de exam para cambiar 'exam_number' por 'number'
- Se cambiaron las validaciones de StoreExamRequest para que valide teacher_id y subject_id por separado
- Store y Update ahora verifican si existe un teacher_subject_id del teacher_id y subject_id ingresados en el formulario, si existe, guarda su id, sino existe, crea un nuevo registro con estos datos y lo guarda.
- Se creó un modelo TeacherSubject el cual está relacionado con Teacher y Subject
- En el modelo de Exam se crearon las relaciones de Teacher y Subject

MaksiConGit committed:
feat: CRUD Exam

- Se realizó el CRUD de Exam con las validaciones.
- Se creó la ruta
- Se cambió  la seeder, factory y migration de exam para cambiar 'exam_number' por 'number'
- Se cambiaron las validaciones de StoreExamRequest para que valide teacher_id y subject_id por separado
- Store y Update ahora verifican si existe un teacher_subject_id del teacher_id y subject_id ingresados en el formulario, si existe, guarda su id, sino existe, crea un nuevo registro con estos datos y lo guarda.
- Se creó un modelo TeacherSubject el cual está relacionado con Teacher y Subject
- En el modelo de Exam se crearon las relaciones de Teacher y Subject

LucasDLabo commented:
- El input de fecha en el Create y Edit debe ser un tipo date.
- El formulario de edición no mantiene los datos actuales que tiene el registro.
- Si se elimina un profesor, a la hora de editar un examen de ese profesor, la página da error.
- Si se elimina una materia, toda la pagina index da error.
- En el show dice, "Exámenes Eliminadas". 🤬

> Nota: podría existir una función que si se elige o una materia o un profesor primero, el siguiente campo del que no se eligió, traiga o todas sus materias o todos sus profesores relacionados. Si elijo la materia primero, que me traiga solamente los profesores que tengan esa materia. Si elijo el profesor, que me traiga todas las materias de ese profesor. ¿No debería dejarme asignar solamente un examen de un profesor que si tiene la materia y viceversa? (Es decir que habría que separar la función de crear la relación teacher_subject del create exams) (Además que cuando se edita un examen, la relación previamente creada no se borra ni se sobrescribe ¿Intencional?) 
¿Posible nueva historia de usuario para generar las relaciones de teacher_subject aparte?

MaksiConGit committed:
Merge branch '136-exam-model-route-controller' of https://github.com/MaksiConGit/proyecto-final-tercero into 136-exam-model-route-controller

MaksiConGit committed:
fix: CRUD Exam v1.0

- Input de fechas cambiado de tipo text a date
- Formulario de edición ahora muestra los datos actuales del registro (aunque creo que hay un error el cual hace que se repita el registro que ya posee)
- Ahora se maneja el error al eliminar un profesor en edit, el valor del profesor se reemplaza por 'No asignado'
- Si se elimina una materia, se eliminan los exámenes de esa materia y se especifica que su materia está eliminada (las eliminaciones de los exámenes se hace en el destroy de subject, antes de que se elimine el subject)
- El mensaje de "Exámenes Eliminadas" fue arreglado
- Se añaden relaciones de TeacherSubject y Subject

LucasDLabo commented:
- Si elimino un profesor que previamente tenia asignado un examen, a la hora de editar ese examen para asignarle un profesor nuevo, la página da error.
- Colocar a todos los campos que tengan $teacher->name, la columna 'lastname', para poder ver el nombre completo del profesor. Si tengo un profesor que se llama Nicolás Rotili, en los labels solo aparece "Nicolás"
- Corregir el duplicado de profesores dentro del select en edit examenes con este código 👇
```
@foreach ($teachers as $teacher)
    <option value="{{ $teacher->id }}"
        {{ old('teacher_id', $exam->teacherSubject->teacher->id) == $teacher->id ? 'selected' : '' }}>
        {{ $teacher->name . " " . $teacher->lastname }}
    </option>
@endforeach
```
- Corregir el duplicado de materias dentro del select en edit examenes con este código 👇
```
@foreach ($subjects as $subject)
    <option value="{{ $subject->id }}"
        {{ old('subject_id', $exam->teacherSubject->subject->id) == $subject->id ? 'selected' : '' }}>
        {{ $subject->name }}
    </option>
@endforeach
```

MaksiConGit committed:
fix: CRUD Exam v1.1

- Corregido duplicados al seleccionar materias y profesores
- Se ve el nombre y apellido del profesor en todo momento
- El error sobre editar el profesor al eliminarlo era por un fallo en el código duplicado
 - La opción "Seleccione una materia" y "Seleccione un profesor" está predefinida si no tiene una materia o un profesor, además está deshabilitada para no poder selccionarse

LucasDLabo commented:
Perfecto, confió en que el error no sucede más 🤓. Me gusto lo del disabled ✌.

LucasDLabo reviewed (dismissed):
Perfecto, confió en que el error no sucede más 🤓. Me gusto lo del disabled ✌.
